lang/go/bindings/cne/lport.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	pcfg.umem_addr = (*C.char)(C.mmap_addr(unsafe.Pointer(ud.mm)))
+	pcfg.umem_size = C.ulong(C.mmap_size(unsafe.Pointer(ud.mm), nil, nil))
+	pcfg.busy_timeout = C.ushort(lportInfo.BusyTimeout)
+	pcfg.busy_budget = C.ushort(lportInfo.BusyBudget)
+	pcfg.flags = C.ushort(lportInfo.LPortFlags())
+
+	if lportInfo.Region >= len(ud.rInfo) {
+		return fmt.Errorf("lport region %d in umem %s for lport %s is not configured",
+			lportInfo.Region, lportInfo.Umem, lportName)
+	}
+
+	pcfg.addr = unsafe.Pointer(ud.rInfo[lportInfo.Region].addr)
+	if pcfg.addr == nil {
+		return fmt.Errorf("lport %s umem %s region index %d >= %d or not configured correctly",
+			lportName, lportInfo.Umem, lportInfo.Region, len(ud.rInfo))
+	}
+
+	pcfg.bufcnt = C.uint(ud.rInfo[lportInfo.Region].bufCnt)
+
+	pcfg.pi = ud.rInfo[lportInfo.Region].pool
+
+	lportId := int(C.pktdev_port_setup(pcfg))
+	if lportId < 0 {
+		return fmt.Errorf("pktdev_port_setup() failed for lport %s", lportName)
+	}
+
+	// Build up the LPort structure only once per LPortInfo structure
+	lport := newLPort(lportName, netdevName, lportId)
+
+	// Add the Lport to the LPort slice and map
+	lportList[lportId] = lport
+	lportMap[lportName] = lport
+
+	return nil
+}
+
+// setupLPorts sets up all LPorts and initializes them
+func (sys *System) setupLPorts() error {
+
+	var lportNames []string
+
+	lportList = make([]*LPort, len(sys.jcfg.LPortInfoMap()))
+
+	// Force the LPort names to be sorted to force order on the LPortInfo structures
+	for lportName := range sys.jcfg.LPortInfoMap() {
+		lportNames = append(lportNames, lportName)
+	}
+
+	sort.Slice(lportNames, func(i, j int) bool { return lportNames[i] < lportNames[j] })
+
+	for _, lportName := range lportNames {
+
+		lportInfo, ok := sys.jcfg.LPortInfoMap()[lportName]
+		if !ok {
+			return fmt.Errorf("LPort %s not found in LPortMap", lportName)
+		}
+
+		if err := sys.setupLPort(lportName, lportInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// cleanupLPorts will cleanup all of the LPorts
+func (sys *System) cleanupLPorts() error {
+
+	for _, lport := range lportList {
+		if lport.lportId < 0xFFFF {
+			ret := C.pktdev_close(C.ushort(lport.lportId))
+			if ret < 0 {
+				return fmt.Errorf("error pktdev_close for lport %s", lport.name)
+			}
+		}
+	}
+	return nil
+}
+
+// LPortMap returns the MAP of LPorts
+func (sys *System) LPortMap() map[string]*LPort {
+
+	return lportMap
+}
+
+// LPortList returns the List of LPorts
+func (sys *System) LPortList() []*LPort {
+
+	return lportList
+}
+
+// LPortByName returns the LPort for the given name
+func (sys *System) LPortByName(name string) *LPort {
+
+	lport, ok := lportMap[name]
+	if !ok {
+		fmt.Printf("LPort %#v not found\n", name)
+		return nil
+	}
+	return lport
+}
+
+// LPortsByName return a list of LPort structure pointers
+func (sys *System) LPortsByName(names []string) []*LPort {
+
+	var lports []*LPort
+
+	for _, name := range names {
+		lport := sys.LPortByName(name)
```

Comment:
just a style nit... in some places we do an initialisation like here then follow up with the nil check... then in other places we do the one liner variation like:

```
	if ret := C.pktdev_stats_get(p.lportId, &stats); ret < 0 {
		return nil, fmt.Errorf("LPortStats failed with error code %d", ret)
	}
```

Maybe it would be nice to pick a style... it's not a show stopper/blocker - just a nit

Reply:
Let me look at them, I did try to fix up the ones I found to be the same when able to.